src/test/java/org/apache/commons/collections4/bloomfilter/BitMapProducerFromWrappedBloomFilterTest.java

```diff
@@ -23,13 +23,25 @@ public class BitMapProducerFromWrappedBloomFilterTest extends AbstractBitMapProd
     @Override
     protected BitMapProducer createEmptyProducer() {
         return new WrappedBloomFilter(new DefaultBloomFilterTest.SparseDefaultBloomFilter(shape)) {
+            @Override
```

Comment:
I'd like to review the API design of the `BloomFilter#merge()` method. It seems that (after a quick look) it always returns `true`. What's a user to do with a `false` result? Would throwing an exception be better?
If that's the case I think it _might_ be nicer to return `this` from a merge which would allow to use the merge as a fluent API call. For example, this copy method would become 
`return new DefaultBloomFilterTest.SparseDefaultBloomFilter(shape).merge(wrapped);`

Reply:
The `boolean` return was added to support the `CountingBloomFilter`. That filter can be invalidated by a merge operation. If we remove the `boolean` return, then a separate call is required for each merge to check that the filter is still valid.

In 99.9999% of cases this will be true as the current implementation can count up to `Integer.MAX_VALUE`. A more space efficient counting Bloom filter would use a smaller primitive type to store the counts such as a `byte`. This would only be able to count up to 127 merge operations at the same cell index while retaining the sign bit to check for overflow and invalidation of the filter.

This does require a bit of consideration. The `CountingBloomFilter` can **add** and **remove** Bloom filters. However it should also have use as a regular Bloom filter and be able to check for presence of an object with a given false-positive rate. So a future improvement (not yet implemented) is to consider the `Shape` when constructing the filter and use the most efficient storage.

For example, if a filter is created that will accept up to 30000 items before saturation (and there is a decline in the false positive rate beyond the target value), then the counts could be maintained using a `short`.

Note that I am not opposed to an API change here. The `CountingBloomFilter` is a specialised case and usage of it requires some knowledge of its limitations. So any API change to make the `BloomFilter` more fluent should consider how to document that the merge can fail in some cases.

Reply:
Ty for the explanation. We can leave it as is then. If there is any way to improve the Javadoc to help non-bloom filter experts, then that would be nice.